test/core/event_engine/posix/poller_fork_test.cc

```diff
@@ -0,0 +1,527 @@
+  // be adjusted in the future!
+  MutableSlice slice = MutableSlice::CreateUninitialized(4 * 1024 * 1024);
+  std::fill(slice.begin(), slice.end(), 42);
+  write_buffer.Append(Slice(std::move(slice)));
+  ASSERT_FALSE(endpoint->Read(read_status.Setter(), &read_buffer, nullptr));
+  ASSERT_FALSE(endpoint->Write(write_status.Setter(), &write_buffer, nullptr))
+      << "Need to send more data";
+  LOG(INFO) << "Before fork in child";
+  // Let the data reach the buffers
+  absl::SleepFor(absl::Milliseconds(50 * grpc_test_slowdown_factor()));
+  ee()->fork_support_for_tests()->BeforeFork();
+  ee()->fork_support_for_tests()->AfterFork(true);
+  LOG(INFO) << "After fork in child";
+  EXPECT_THAT(read_status.AwaitStatus(),
+              StatusIs(absl::StatusCode::kResourceExhausted));
+  EXPECT_THAT(write_status.AwaitStatus(),
+              StatusIs(absl::StatusCode::kResourceExhausted));
+  // Starting read and write post-fork will fail asynchronously and return the
+  // status.
+  ASSERT_FALSE(endpoint->Read(read_status.Setter(), &read_buffer, nullptr));
+  ASSERT_FALSE(endpoint->Write(write_status.Setter(), &write_buffer, nullptr));
+  EXPECT_THAT(read_status.AwaitStatus(), StatusIs(absl::StatusCode::kInternal));
+  EXPECT_THAT(write_status.AwaitStatus(), ::testing::Not(IsOk()));
+}
+
+class TestScheduler {
+ public:
+  static std::shared_ptr<TestScheduler> Start() {
+    auto scheduler = std::make_shared<TestScheduler>();
+    scheduler->Start(scheduler);
+    return scheduler;
+  }
+
+  ~TestScheduler() {
+    {
+      grpc_core::MutexLock lock(&mu_);
+      done_ = true;
+      cond_.SignalAll();
+    }
+    if (background_thread_.joinable()) {
+      background_thread_.join();
+    }
+  }
+
+  void Schedule(absl::AnyInvocable<void() const> op)
+      ABSL_NO_THREAD_SAFETY_ANALYSIS {
+    if (is_in_scheduler) {
+      pending_.push(std::move(op));
+    } else {
+      grpc_core::MutexLock lock(&mu_);
+      pending_.push(std::move(op));
+    }
+    cond_.SignalAll();
+  }
+
+ private:
+  void Start(std::shared_ptr<TestScheduler> scheduler) {
+    background_thread_ = std::thread(
+        [scheduler = std::move(scheduler)]() { scheduler->EventQueue(); });
+  }
+
+  void EventQueue() {
+    is_in_scheduler = true;
+    absl::Cleanup reset_is_in_scheduler = []() { is_in_scheduler = false; };
+    grpc_core::MutexLock lock(&mu_);
+    while (!done_) {
+      while (!pending_.empty()) {
+        pending_.front()();
+        pending_.pop();
+      }
+      cond_.Wait(&mu_);
+    }
+  }
+
+  std::thread background_thread_;
+  grpc_core::Mutex mu_;
+  grpc_core::CondVar cond_;
+  std::queue<absl::AnyInvocable<void() const>> pending_ ABSL_GUARDED_BY(&mu_);
+  bool done_ ABSL_GUARDED_BY(&mu_) = false;
+};
+
+class PollCycle {
+ public:
+  explicit PollCycle(std::shared_ptr<TestScheduler> scheduler,
+                     absl::AnyInvocable<void() const> op)
+      : scheduler_(std::move(scheduler)), op_(std::move(op)), is_scheduled_(1) {
```

Comment:
This initializes `is_scheduled_` to 1, but we're initializing it to 0 where it's declared below.  Let's just initialize it to 1 below, so we don't need to mention it here.